images/git-remote-sync/sync-git-remotes.py

```diff
@@ -0,0 +1,116 @@
+import argparse
+import base64
+import os
+import re
+import subprocess
+import sys
+import tempfile
+from contextlib import contextmanager
+
+try:
+    import sentry_sdk
+    sentry_sdk.init(
+        # This cron job only runs once weekly,
+        # so just record all transactions.
+        traces_sample_rate=1.0,
+    )
+except Exception:
+    print("Could not configure sentry.")
+
+
+# Copied from SpackCIBridge.py
+def setup_ssh(self, ssh_key_base64):
+    """Start the ssh agent."""
+    print("Starting ssh-agent")
+    output = subprocess.run(["ssh-agent", "-s"], check=True, stdout=subprocess.PIPE).stdout
+
+    # Search for PID in output.
+    pid_regexp = re.compile(r"SSH_AGENT_PID=([0-9]+)")
+    match = pid_regexp.search(output.decode("utf-8"))
+    if match is None:
+        print("WARNING: could not detect ssh-agent PID.", file=sys.stderr)
+        print("ssh-agent will not be killed upon program termination", file=sys.stderr)
+    else:
+        pid = match.group(1)
+        os.environ["SSH_AGENT_PID"] = pid
+        self.cleanup_ssh_agent = True
+
+    # Search for socket in output.
+    socket_regexp = re.compile(r"SSH_AUTH_SOCK=([^;]+);")
+    match = socket_regexp.search(output.decode("utf-8"))
+    if match is None:
+        print("WARNING: could not detect ssh-agent socket.", file=sys.stderr)
+        print("Key will be added to caller's ssh-agent (if any)", file=sys.stderr)
+    else:
+        socket = match.group(1)
+        os.environ["SSH_AUTH_SOCK"] = socket
+
+    # Add the key.
+    ssh_key = base64.b64decode(ssh_key_base64)
+    ssh_key = ssh_key.replace(b"\r", b"")
+    with tempfile.NamedTemporaryFile() as fp:
+        fp.write(ssh_key)
+        fp.seek(0)
+        subprocess.run(["ssh-add", fp.name], check=True)
+
+
+def setup_repo(pull_remote, push_remote, branch):
+    subprocess.run(["git", "init"], check=True)
+    subprocess.run(["git", "config", "user.email", "noreply@spack.io"], check=True)
+    subprocess.run(["git", "config", "user.name", "spackbot"], check=True)
+    subprocess.run(["git", "config", "advice.detachedHead", "false"], check=True)
+
+    subprocess.run(["git", "remote", "add", "origin", pull_remote], check=True)
+    subprocess.run(["git", "remote", "set-url", "--push", "origin", push_remote], check=True)
+
+    fetch_args = ["git", "fetch", "-q", "origin", branch]
+    subprocess.run(fetch_args, check=True, stdout=subprocess.PIPE).stdout
+    subprocess.run(["git", "checkout", branch], check=True)
+
+
+@contextmanager
+def chworkingdir(path):
```

Comment:
Did you look at the [one](https://docs.python.org/3/library/contextlib.html#contextlib.chdir) provided by `contextlib`?  Maybe you don't need to write your own.

Reply:
It would have to be something like `with tempdir.TemporaryDirectory() as workingdir, contextmanager.chdir(workingdir)` which I didn't like. The one I have lets be choose whether to chdir to a temp dir or a persistent dir which was nice to test this locally.

Reply:
I did the something similar for testing protected publish locally, [this](https://github.com/scottwittenburg/spack-infrastructure/blob/protected-publish/images/protected-publish/publish.py#L444-L448) is how I avoided forcing use of a temporary directory.

Reply:
I see, the way yield works is still a bit of magic to me, I didn't know I could construct it, return the yielded result, and use it in a with clause. 🪄